tests/Controller/AdminProductVariantAjaxTest.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Tests\Controller;
+
+use ApiTestCase\JsonApiTestCase;
+use Symfony\Component\BrowserKit\Cookie;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken;
+
+final class AdminProductVariantAjaxTest extends JsonApiTestCase
+{
+    /** @test */
+    public function it_denies_access_to_product_variants_for_not_authenticated_user(): void
+    {
+        $this->client->request('GET', '/admin/ajax/product-variants/search-all');
+
+        $response = $this->client->getResponse();
+
+        $this->assertEquals($response->getStatusCode(), Response::HTTP_FOUND);
+    }
+
+    /** @test */
+    public function it_returns_only_specified_part_of_all_product_variants_for_empty_phrase(): void
+    {
+        $this->loadFixturesFromFile('authentication/administrator.yml');
+        $this->loadFixturesFromFiles(['resources/product_variants.yml']);
+
+        $this->authenticateAdminUser();
+
+        $this->client->request('GET', '/admin/ajax/product-variants/search-all?phrase=&limit=25');
+
+        $response = $this->client->getResponse();
+
+        $this->assertResponse($response, 'ajax/product_variant/index_response', Response::HTTP_OK);
+    }
+
+    /** @test */
+    public function it_throws_type_error_when_phrase_is_not_specified(): void
+    {
+        $this->loadFixturesFromFile('authentication/administrator.yml');
+        $this->loadFixturesFromFiles(['resources/product_variants.yml']);
+
+        $this->authenticateAdminUser();
+
+        $this->expectException(\TypeError::class);
```

Comment:
I think this is better handled on the route configuration level, so it throws an exception instead of a TypeError.
The TypeError can come from somewhere else.
Something like this:
```yaml
    requirements:
        phrase: '.+'
```

Reply:
Sadly with doesn't work with query parameters...

Reply:
I would rather add a default value of an empty string for phrase instead of expecting a TypeError.
High level logic like this should not rely on the internal PHP errors.
Moreover, in production environments this will result in a 500 Internal Server Error, which is not good.

Reply:
AFAIK any unhandled throwable will result 500 in production because there is no fallback throwable listener in Sylius.
I tried adding `requirements` but it didn't work for me. Just like in case of your `limit` in `defaults` (I tested it as well and it didn't work) it's only work for slug.
I have no idea how to control the exception throwing instead of type error.

Reply:
Yes, requirements don't work for query parameters.
And the default won't work for the same reason as the limit.
How about we use a similar solution then:
```
phrase: expr:service('request_stack').getCurrentRequest().query.get('limit', '')
```

Reply:
This way there won't be an error at all, it will fallback to an empty string.